### PR TITLE
Clock went backwards corruption check

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
TimeLock should validates the expected invariant for its timestamp bounds: they must increase with increasing sequence numbers.

**Implementation Description (bullets)**:

- LocalTimestampInvariantsVerifier - verifies  timestamp bounds increase with increasing sequence numbers

- The validation is done batch wise, e.g. [1, n], [n, 2 * n - 1] and so on. Two consecutive batches share
 boundaries, this is done to catch inversion at batch end.

- The verifier uses the existing corruption handling infra i.e. invariant violation will show up in the health check and (when enabled) is capable of killing TimeLock in the case of violation.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added unit tests

**Concerns (what feedback would you like?)**:

- Does the design seem reasonable?

- Can we optimize the implementation?

- Did I miss any edge cases?

**Where should we start reviewing?**:
HistoryAnalyzer.java

**Priority (whenever / two weeks / yesterday)**:
End of week.

### CODE REVIEW

File: timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/HistoryAnalyzer.java [hunk 1]
```diff
@@ -195,4 +206,35 @@ private static int getQuorumSize(List<ConsolidatedLearnerAndAcceptorRecord> reco
...
+    }
+
+    @Value.Immutable
+    interface SequenceAndTimestampPair {
+        @Value.Parameter
+        long sequence();
+
+        @Value.Parameter
+        long timestamp();
+    }
```
Comment (gmaretic) [inline timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/HistoryAnalyzer.java:239]: If you want to be a bit fancy:
```suggestion
        Stream<Long> expectedSortedTimestamps = KeyedStream.stream(record.record())
                .map(val -> getPaxosValueData(val.learnedValue()))
                .filter(Predicates.notNull())
                .mapEntries((sequence, timestamp) -> Maps.immutableEntry(sequence, PtBytes.toLong(timestamp)))
                .entries()
                .sorted(Comparator.comparingLong(Entry::getKey))
                .map(Entry::getValue);
        return StreamEx.of(expectedSortedTimestamps)
                .pairMap((first, second) -> first >= second)
                .anyMatch(x -> x);
```
If you don't want to use `StreamEx`, you can still avoid the immutable by checking `expectedSortedTimestamps` in a for loop, but just compare neighbours instead of maintaining a global maximum

**Review by gmaretic — COMMENTED** [review]
Makes sense, though the analyzer itself need some polish

**Comment by sudiksha27** [comment]
This will need redesigning. Will be moving the check out of History analyzer and adding a separate thread to verify timestamp invariants.

File: timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifier.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.sql.DataSource;
+import one.util.streamex.StreamEx;
+
+public class LocalTimestampInvariantsVerifier {
+    @VisibleForTesting
+    static final int DELTA = 5;
+
+    private final SqlitePaxosStateLogHistory sqlitePaxosStateLogHistory;
```
Comment (sudiksha27) [inline timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifier.java:45]: This won't work, need to fetch the next seq number > upper and use range [lower, next_seq_after_upper] for analysis.

File: timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalCorruptionDetector.java [hunk 3]
```diff
@@ -34,30 +34,44 @@
...
     private void scheduleWithFixedDelay() {
         executor.scheduleWithFixedDelay(
                 () -> {
-                    localCorruptionReport = analyzeHistoryAndBuildCorruptionHealthReport();
+                    CorruptionHealthReport paxosRoundCorruptionReport = analyzeHistoryAndBuildCorruptionHealthReport();
+                    CorruptionHealthReport timestampInvariantsReport =
+                            timestampInvariantsVerifier.timestampInvariantsHealthReport();
+                    localCorruptionReport = ImmutableCorruptionHealthReport.builder()
+                            .putAllViolatingStatusesToNamespaceAndUseCase(
+                                    paxosRoundCorruptionReport.violatingStatusesToNamespaceAndUseCase())
```
Comment (gmaretic) [inline timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalCorruptionDetector.java:71]:
```suggestion
                                                        .from(paxosRoundCorruptionReport)
```

File: timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifier.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                .collectToSetMultimap();
+        return ImmutableCorruptionHealthReport.builder()
+                .violatingStatusesToNamespaceAndUseCase(namespacesExhibitingViolations)
+                .build();
+    }
+
+    private CorruptionCheckViolation timestampInvariantsViolationLevel(NamespaceAndUseCase namespaceAndUseCase) {
+        Stream<Long> expectedSortedTimestamps = KeyedStream.stream(getLearnerLogs(namespaceAndUseCase))
+                .map(PaxosValue::getData)
+                .filter(Predicates.notNull())
```
Comment (gmaretic) [inline timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifier.java:75]:
```suggestion
                .filter(Objects::nonNull)
```

File: timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifier.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+ * This class validates that timestamp bounds increase with increasing sequence numbers.
+ *
+ * The validation is done batch wise, e.g. [1, n], [n, 2 * n - 1] and so on. Two consecutive batches share
+ * boundaries, this is expected and is done to catch inversion at batch end.
+ * */
+public class LocalTimestampInvariantsVerifier {
+    @VisibleForTesting
+    public static final int LEARNER_LOG_BATCH_SIZE_LIMIT = 250;
+
+    public static final long MIN_SEQUENCE_TO_BE_VERIFIED = -1L;
```
Comment (gmaretic) [inline timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifier.java:51]:
```suggestion
    public static final long MIN_SEQUENCE_TO_BE_VERIFIED = Long.MIN_VALUE;
```

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import org.junit.Test;
+
+public class LocalTimestampInvariantsVerifierTest {
+    @Rule
+    public TimeLockCorruptionDetectionHelper helper = new TimeLockCorruptionDetectionHelper();
+
+    @Test
+    public void detectCorruptionIfClockWentBackwardsOnNode() {
+        helper.writeLogsOnDefaultLocalServer(1, LEARNER_LOG_BATCH_SIZE_LIMIT - 1);
+        helper.forceTimestampToGoBackwards(5);
```
Comment (gmaretic) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java:31]: This method name is kinda misleading, it's actually forcing the timestamp to go forward / sequence to go backwards. It's only making the timestamp go backwards for the sequence after, right?

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    public void detectIfClockWentBackwardsAtBatchStart() {
+        helper.writeLogsOnDefaultLocalServer(1, LEARNER_LOG_BATCH_SIZE_LIMIT);
+        helper.forceTimestampToGoBackwards(1);
+        helper.assertClockWentBackwards();
+    }
+
+    @Test
+    public void detectIfClockWentBackwardsAtBatchEnd() {
+        helper.writeLogsOnDefaultLocalServer(1, LEARNER_LOG_BATCH_SIZE_LIMIT);
+        helper.forceTimestampToGoBackwards(LEARNER_LOG_BATCH_SIZE_LIMIT / 2);
```
Comment (gmaretic) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java:45]: Shouldn't this then be `helper.forceTimestampToGoBackwards(LEARNER_LOG_BATCH_SIZE_LIMIT - 1);`?

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test
+    public void resetsProgressIfNotEnoughLogsForVerification() {
+        helper.writeLogsOnDefaultLocalServer(1, 1);
+        // No signs of corruption
+        helper.assertLocalTimestampInvariantsStand();
+
+        helper.writeLogsOnDefaultLocalServer(2, LEARNER_LOG_BATCH_SIZE_LIMIT);
+        helper.forceTimestampToGoBackwards(LEARNER_LOG_BATCH_SIZE_LIMIT / 2);
```
Comment (gmaretic) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java:80]:
```suggestion
        helper.forceTimestampToGoBackwards(1);
```
This way we test that we don't skip the one entry from first check. Also I guess do this before the first check to be consistent

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java [hunk 9]
```diff
@@ -105,6 +105,31 @@ void assertViolationDetectedForNamespaceAndUseCases(
         return timeLockCorruptionTestSetup.createStatLogForNamespaceAndUseCase(namespaceAndUseCase);
     }
 
+    void assertClockWentBackwards() {
```
Comment (gmaretic) [inline timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/TimeLockCorruptionDetectionHelper.java:108]: Maybe rename to `assertClockWentBackwardsInNextBatch` to be clearer? Same for the method below

**Review by gmaretic — COMMENTED** [review]
Mostly looks good, have some nits

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java (hunk 6 above)
Comment (sudiksha27) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java]: Ah yes, that makes sense. Basically what I mean is - create inversion b/w this sequence and the next. I have changed the name to createTimestampInversion now.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java (hunk 7 above)
Comment (sudiksha27) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java]: Good catch! yes, this is a copy pasting mistake.

File: timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java (hunk 8 above)
Comment (sudiksha27) [reply timelock-corruption-detection/src/test/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifierTest.java]: Done.

**Review by gmaretic — APPROVED** [review]
LGTM
